db/rdbms/src/main/java/io/camunda/db/rdbms/write/service/FlowNodeInstanceWriter.java

```diff
@@ -36,13 +38,47 @@ public void create(final FlowNodeInstanceDbModel flowNode) {
             flowNode));
   }
 
-  public void end(final long flowNodeKey, final FlowNodeState state, final OffsetDateTime endDate) {
-    final var dto = new EndFlowNodeDto(flowNodeKey, state, endDate);
-    executionQueue.executeInQueue(
-        new QueueItem(
-            ContextType.FLOW_NODE,
-            flowNodeKey,
-            "io.camunda.db.rdbms.sql.FlowNodeInstanceMapper.updateStateAndEndDate",
-            dto));
+  public void end(
+      final long flowNodeInstanceKey, final FlowNodeState state, final OffsetDateTime endDate) {
+    final var dto = new EndFlowNodeDto(flowNodeInstanceKey, state, endDate);
+    final boolean wasMerged =
+        executionQueue.tryMergeWithExistingQueueItem(new EndFlowNodeToInsertMerger(dto));
+
+    if (!wasMerged) {
+      executionQueue.executeInQueue(
+          new QueueItem(
+              ContextType.FLOW_NODE,
+              flowNodeInstanceKey,
+              "io.camunda.db.rdbms.sql.FlowNodeInstanceMapper.updateStateAndEndDate",
+              dto));
+    }
+  }
+
+  public static class EndFlowNodeToInsertMerger implements QueueItemMerger {
+
+    private final EndFlowNodeDto dto;
+
+    public EndFlowNodeToInsertMerger(final EndFlowNodeDto dto) {
+      this.dto = dto;
+    }
+
+    @Override
+    public boolean canBeMerged(final QueueItem queueItem) {
+      return queueItem.contextType() == ContextType.FLOW_NODE
+          && queueItem.id().equals(dto.flowNodeInstanceKey())
+          && queueItem.parameter() instanceof FlowNodeInstanceDbModel;
```

Comment:
Should always be a FlowNodeInstanceDbModel, or? Since instanceof is always a little bit expansive when it comes to performance, we could maybe remove it?

Reply:
same for the other merger

Reply:
Currently yes. But in future maybe we also have other parameter objects (e.g. when adding incidents). Then we could either include the statement type to the condition or check for the instance type of the parameter. I've for now chosen the instance type to avoid magic string constants.